pkg/hubble/relay/pool/manager.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pool
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	peerpb "github.com/cilium/cilium/api/v1/peer"
+	hubblePeer "github.com/cilium/cilium/pkg/hubble/peer"
+	"github.com/cilium/cilium/pkg/lock"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/connectivity"
+)
+
+// PeerManager defines the functions a peer manager must implement when
+// handling peers and respective connections.
+type PeerManager interface {
+	// Start instructs the manager to start peer change notification handling
+	// and connection management.
+	Start()
+	// Stop stops any peer manager activity.
+	Stop()
+	// List returns a list of peers with active connections. If a peer cannot
+	// be connected to; its Conn attribute must be nil.
+	List() []Peer
+	// ReportOffline allows the caller to report a peer as being offline. The
+	// peer is identified by its name.
+	ReportOffline(name string)
+}
+
+// Peer is like peer.Peer but includes a Conn attribute to reach the peer's
+// gRPC API endpoint.
+type Peer struct {
+	hubblePeer.Peer
+	Conn ClientConn
+}
+
+type peer struct {
+	mu lock.Mutex
+	hubblePeer.Peer
+	conn            ClientConn
+	connAttempts    int
+	nextConnAttempt time.Time
+}
+
+// Manager implements the PeerManager interface.
+type Manager struct {
```

Comment:
Any particular reason `Manager` is public? My understanding is that exposing the `PeerManager` interface should be enough.

Reply:
`Manager` is an implementation of the `PeerManager` interface and it needs to be public because this is what the constructor returns. See also [CodeReviewComments#interfaces](https://github.com/golang/go/wiki/CodeReviewComments#interfaces):
> The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.